File: 67. Add Binary/addBinary.java
class Solution {
    public String addBinary(String a, String b) {
        int carry = 0;
        char[] arr1 = a.toCharArray();
        char[] arr2 = b.toCharArray();
        int p1 = arr1.length - 1;
        int p2 = arr2.length - 1;
        int p = Math.max(p1, p2) + 1;
        char[] s = new char[p + 1];
        while (p1 >= 0 && p2 >= 0) {
            int x = arr1[p1] - '0';
            int y = arr2[p2] - '0';
            int sum = x + y + carry;
            if (sum > 1) {
                carry = 1;
                s[p] = (char)('0' + (sum % 2));
            } else {
                carry = 0;
                s[p] = (char)('0' + sum);
            }
            p1--;
            p2--;
            p--;
        }
        while (p1 >= 0) {
            int x = arr1[p1] - '0';
            int sum = x + carry;
            if (sum > 1) {
                carry = 1;
                s[p] = (char)('0' + (sum % 2));
            } else {
                carry = 0;
                s[p] = (char)('0' + sum);
            }
            p1--;
            p--;
        }
        while (p2 >= 0) {
            int x = arr2[p2] - '0';
            int sum = x + carry;
            if (sum > 1) {
                carry = 1;
                s[p] = (char)('0' + (sum % 2));
            } else {
                carry = 0;
                s[p] = (char)('0' + sum);
            }
            p2--;
            p--;
        }
        if (carry > 0) {
            s[p] = (char)('0' + carry);
            return String.valueOf(s);
        } else {
            return String.valueOf(Arrays.copyOfRange(s, 1, s.length));
        }

    }
}

File: 67. Add Binary/addBinary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        arr1 = [(ord(i) - ord('0')) for i in a]
        arr2 = [(ord(i) - ord('0')) for i in b]
        if len(arr1) < len(arr2):
            arr1 = [0] * (len(arr2) - len(arr1)) + arr1
        if len(arr2) < len(arr1):
            arr2 = [0] * (len(arr1) - len(arr2)) + arr2
        
        n = len(arr1) + 1
        res = [''] * n
        carry = 0

        for i in range(n - 2, -1, -1):
            x = arr1[i]
            y = arr2[i]
            sum = x + y + carry
            if sum > 1:
                carry = 1
                res[i + 1] = chr(ord('0') + sum % 2);
            else:
                carry = 0
                res[i + 1] = chr(ord('0') + sum)


        if carry > 0:
            res[0] = chr(ord('0') + carry)
        return ''.join(res)
